src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs

```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+            => new FunctionPointerMethodSymbol(
+                this.CallingConvention,
+                this.RefKind,
+                substitutedReturnType,
+                this.RefCustomModifiers,
+                this.Parameters,
+                substitutedParameterTypes);
+
+        internal FunctionPointerMethodSymbol MergeEquivalentTypes(FunctionPointerMethodSymbol signature, VarianceKind variance)
+        {
+            var mergedParameterTypes = ImmutableArray<TypeWithAnnotations>.Empty;
+            bool hasParamChanges = false;
+            if (_parameters.Length > 0)
+            {
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(_parameters.Length);
+                var paramVariance = variance switch
+                {
+                    VarianceKind.In => VarianceKind.Out,
+                    VarianceKind.Out => VarianceKind.In,
+                    _ => VarianceKind.None,
+                };
+
+                for (int i = 0; i < _parameters.Length; i++)
+                {
+                    var mergedParameterType = _parameters[i].TypeWithAnnotations.MergeEquivalentTypes(signature._parameters[i].TypeWithAnnotations, paramVariance);
+                    builder.Add(mergedParameterType);
+                    if (!mergedParameterType.IsSameAs(_parameters[i].TypeWithAnnotations))
+                    {
+                        hasParamChanges = true;
+                    }
+                }
+
+                mergedParameterTypes = builder.ToImmutableAndFree();
+            }
+
+            var mergedReturnType = ReturnTypeWithAnnotations.MergeEquivalentTypes(signature.ReturnTypeWithAnnotations, variance);
+            if (hasParamChanges || !mergedReturnType.IsSameAs(ReturnTypeWithAnnotations))
+            {
+                return SubstiteParameterSymbols(mergedReturnType, mergedParameterTypes);
+            }
+            else
+            {
+                return this;
+            }
+        }
+
+        public FunctionPointerMethodSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        {
+            var transformedParameterTypes = ImmutableArray<TypeWithAnnotations>.Empty;
+            bool hasParamChanges = false;
+            if (_parameters.Length > 0)
+            {
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(_parameters.Length);
+                foreach (var param in _parameters)
+                {
+                    var transformedType = transform(param.TypeWithAnnotations);
+                    builder.Add(transformedType);
+                    if (!transformedType.IsSameAs(param.TypeWithAnnotations))
+                    {
+                        hasParamChanges = true;
+                    }
+                }
+
+                transformedParameterTypes = builder.ToImmutableAndFree();
+            }
+
+            var transformedReturn = transform(ReturnTypeWithAnnotations);
```

Comment:
>var transformedReturn = transform(ReturnTypeWithAnnotations); [](start = 12, length = 61)

I think this violates our general approach that transforms are applied to types in the order in which types appear in metadata signature. I believe, in metadata signature return type comes before parameter types.

Reply:
Please add targeted tests nullable encoding/decoding for function pointers in metadata.

---
In reply to: [406858800](https://github.com/dotnet/roslyn/pull/43041#discussion_r406858800) [](ancestors = 406858800)

Reply:
Nullable tests are already covered by a prototype comment and are out of scope for this PR. Will add this to the list to test when that is addressed.

---
In reply to: [406859222](https://github.com/dotnet/roslyn/pull/43041#discussion_r406859222) [](ancestors = 406859222,406858800)

Reply:
It is fine to postpone testing, but implementation should be fixed

---
In reply to: [406862984](https://github.com/dotnet/roslyn/pull/43041#discussion_r406862984) [](ancestors = 406862984,406859222,406858800)